src/logic/wallets/getWeb3.ts

```diff
@@ -74,18 +75,47 @@ export const getChainIdFrom = (web3Provider: Web3): Promise<number> => {
   return web3Provider.eth.getChainId()
 }
 
+export const isHardwareWallet = (wallet: Wallet): boolean => {
+  const isSupportedHardwareWallet = [WALLET_PROVIDER.LEDGER, WALLET_PROVIDER.TREZOR].includes(
+    wallet?.name?.toUpperCase() as WALLET_PROVIDER,
+  )
+  const isHardwareWalletType = wallet?.type === 'hardware'
+
+  return isSupportedHardwareWallet || isHardwareWalletType
+}
+
+let isSmartContractWalletCache: Record<ChainId, Record<string, boolean>> = {}
 export const isSmartContractWallet = async (account: string): Promise<boolean> => {
   if (!account) {
     return false
   }
+
+  const chainId = _getChainId()
+
+  if (isSmartContractWalletCache?.[chainId]?.[account] !== undefined) {
+    return isSmartContractWalletCache[chainId][account]
+  }
```

Comment:
Maybe use `memoize` from lodash? We already use in a couple other places.

Reply:
Lodash memoises [based on the argument](https://lodash.com/docs/4.17.15#memoize). If you switch chain, it will return previous-chain's memoized value for the same address.

Reply:
I know, I would define a memoized helper function that takes chainId and address.

Reply:
Richard said that it's cryptographically unlikely that the same address is a contract and an EoA on different chains, so we can simply remove the chainId.

Reply:
I've migrated to lodash' memoization and included a resolver to cover both arguments (as lodash memoizes off the first argument alone). Although unlikely, it's better to be safe than sorry, so I left both in the helper.